repair/sink_source_for_repair.cc

```diff
@@ -0,0 +1,85 @@
+    , _fn(std::move(fn)) {
+}
+
+template<class SinkType, class SourceType>
+void sink_source_for_repair<SinkType, SourceType>::mark_source_closed(unsigned node_idx) {
+    _sources_closed[node_idx] = true;
+}
+
+template<class SinkType, class SourceType>
+future<std::tuple<typename sink_source_for_repair<SinkType, SourceType>::sink_type, typename sink_source_for_repair<SinkType, SourceType>::source_type>>
+sink_source_for_repair<SinkType, SourceType>::get_sink_source(gms::inet_address remote_node, unsigned node_idx) {
+    using value_type = std::tuple<sink_type, source_type>;
+    if (_sinks[node_idx] && _sources[node_idx]) {
+        return make_ready_future<value_type>(value_type(_sinks[node_idx].value(), _sources[node_idx].value()));
+    }
+    if (_sinks[node_idx] || _sources[node_idx]) {
+        return make_exception_future<value_type>(std::runtime_error(format("sink or source is missing for node {}", remote_node)));
+    }
+    return _fn(_repair_meta_id, netw::messaging_service::msg_addr(remote_node)).then_unpack([this, node_idx] (rpc::sink<SinkType> sink, rpc::source<SourceType> source) mutable {
+        _sinks[node_idx].emplace(std::move(sink));
+        _sources[node_idx].emplace(std::move(source));
+        return make_ready_future<value_type>(value_type(_sinks[node_idx].value(), _sources[node_idx].value()));
+    });
+}
+
+template<class SinkType, class SourceType>
+future<> sink_source_for_repair<SinkType, SourceType>::close() {
+    return parallel_for_each(boost::irange(unsigned(0), unsigned(_sources.size())), [this] (unsigned node_idx) mutable {
+        std::optional<rpc::sink<SinkType>>& sink_opt = _sinks[node_idx];
+        auto f = sink_opt ? sink_opt->close() : make_ready_future<>();
+        return f.finally([this, node_idx] {
+            std::optional<rpc::source<SourceType>>& source_opt = _sources[node_idx];
+            if (source_opt && !_sources_closed[node_idx]) {
+                return repeat([&source_opt] () mutable {
+                    // Keep reading source until end of stream
+                    return (*source_opt)().then([] (std::optional<std::tuple<SourceType>> opt) mutable {
+                        if (opt) {
+                            return make_ready_future<stop_iteration>(stop_iteration::no);
+                        } else {
+                            return make_ready_future<stop_iteration>(stop_iteration::yes);
+                        }
+                    }).handle_exception([] (std::exception_ptr ep) {
+                        return make_ready_future<stop_iteration>(stop_iteration::yes);
+                    });
+                });
+            }
+            return make_ready_future<>();
+        });
+    });
+}
+
+template class sink_source_for_repair<repair_stream_cmd, repair_hash_with_cmd>;
+template class sink_source_for_repair<repair_hash_with_cmd, repair_row_on_wire_with_cmd>;
+template class sink_source_for_repair<repair_row_on_wire_with_cmd, repair_stream_cmd>;
```

Comment:
I'm not against this split, but it does raise a question. I assume you split this code out of the main file because it takes a long time of compilation - maybe even minutes.
If this is the case, then maybe (and I'm just guessing here) each of the three instantiations here takes roughly 1/3rd of the compilation time of this file. In this case, can't we *not* make it a template? Can't we use OOP or type erasure (e.g., std::function) so that we have just one of these codes?
I'm not familiar with this sink/source stuff - do we need it to be separately optimized for each of those three types?

Reply:
The main source of complication slowness comes from seastar::rpc. I am not sure how much gain it will give if we get rid of the sink_source_for_repair template class since the main source is rpc::sink and rpc::source template. We need to use it anyway.

Reply:
Can we move *only* the rpc related stuff to a separate cc file? As these are ABIs, these are not changed often so somebody working on the row level repair wouldn't need to incur their compilation cost each time the `row_level.cc` is compiled.

Reply:
@denesb do you mean to left those in the translation unit that is using them - there is only 1 - repair/repair_meta.cc? or something else? not sure I've got what you mean.

Reply:
It was suggested that the largest contributors to compile time are RPC template instantiations. I suggested to move only these to a separate `.cc` file, such that recompiling `row_level.cc` doesn't re-instantiate all the RPC templates. This would speed up incremental compilations.